lib/instrumentation/modules/kafkajs.js

```diff
@@ -0,0 +1,451 @@
+      // https://github.com/elastic/apm/blob/main/specs/agents/tracing-instrumentation-messaging.md#receiving-trace-context
+      // A span link is created from a `traceparent` header in a message.
+      const messages = batch && batch.messages;
+
+      if (messages) {
+        const links = [];
+        const limit = Math.min(
+          messages.length,
+          constants.MAX_MESSAGES_PROCESSED_FOR_TRACE_CONTEXT,
+        );
+
+        for (let i = 0; i < messages.length; i++) {
+          const msg = messages[i];
+          if (msg.headers && msg.headers.traceparent) {
+            links.push({ context: msg.headers.traceparent.toString() });
+
+            if (links.length >= limit) {
+              break;
+            }
+          }
+        }
+        trans._addLinks(links);
+      }
+
+      let result, err;
+      try {
+        result = await origEachBatch.apply(this, arguments);
+      } catch (ex) {
+        // Save the error for use in `finally` below, but re-throw it to
+        // not impact code flow.
+        err = ex;
+        throw ex;
+      } finally {
+        trans.setOutcome(
+          err ? constants.OUTCOME_FAILURE : constants.OUTCOME_SUCCESS,
+        );
+        trans.end();
+      }
+
+      return result;
+    };
+  }
+
+  /**
+   * Returns the patched version of `Kafka.producer` which creates a new
+   * producer with `send` & `sendBatch` methods patched to instrument message sending
+   *
+   * @param {ProducerFactory} origProducer
+   * @returns {ProducerFactory}
+   */
+  function wrapProducer(origProducer) {
+    return function wrappedProducer() {
+      const producer = origProducer.apply(this, arguments);
+
+      shimmer.wrap(producer, 'send', wrapProducerSend);
+      shimmer.wrap(producer, 'sendBatch', wrapProducerSendBatch);
+      return producer;
+    };
+  }
+
+  /**
+   * Returns the instrumented version of `send` which
+   * - creates an exit span each time is called
+   * - propagates trace context through message headers
+   *
+   * @param {Producer['send']} origSend
+   * @returns {Producer['send']}
+   */
+  function wrapProducerSend(origSend) {
+    return async function (record) {
+      const { topic } = record;
+      let span;
+
+      if (!shouldIgnoreTopic(topic, config)) {
+        span = ins.createSpan(
+          `${NAME} send to ${topic}`,
+          TYPE,
+          SUBTYPE,
+          'send',
+          { exitSpan: true },
+        );
+      }
+
+      // W3C trace-context propagation.
+      const runContext = ins.currRunContext();
+      const parentSpan =
+        span || runContext.currSpan() || runContext.currTransaction();
+
+      if (parentSpan) {
+        record.messages.forEach((msg) => {
+          const newHeaders = Object.assign({}, msg.headers);
+          parentSpan.propagateTraceContextHeaders(
+            newHeaders,
+            function (carrier, name, value) {
+              // TODO: why doing it everywhere???
```

Comment:
Because there is a share impl of `propagateTraceContextHeaders` in lib/tracecontext/index.js that is meant to handle bwcompat cases supporting `elastic-apm-traceparent`.  The idea here is that even if `useElasticTraceparentHeader: true` we still don't want to propagate that header for this *new* instrumentation -- only for HTTP instrumentation.  At least in SQS there is a low limit on the number of headers, so adding `elastic-apm-traceparent` that will almost certainly never get used could cause harm.

It certainly does feel silly to have this check everywhere however. Perhaps `.propagateTraceContextHeaders` could grow a `{legacyHeaders: true}` option (or a separate method for this) that would only be called in the HTTP instrumentation. Only that code path would consider adding `elastic-apm-traceparent`.

Reply:
Thanks for the explanation. Makes a lot of sense :)